### PR TITLE
Implement Cluster ID

### PR DESCRIPTION
**Pull request by liyuheng55555**
doc: https://apache-iotdb.feishu.cn/docx/E5hjdaWIwo2JT7xCi6bcvtyinfh

### CODE REVIEW

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/consensus/statemachine/ConfigRegionStateMachine.java [hunk 1]
```diff
@@ -236,6 +236,7 @@ public void notifyLeaderReady() {
     configManager.checkUserPathPrivilege();
     // Add Metric after leader ready
     configManager.addMetrics();
+    configManager.getClusterManager().checkClusterId();
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/consensus/statemachine/ConfigRegionStateMachine.java:239]: async and add comment

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java [hunk 2]
```diff
@@ -203,6 +204,8 @@ public class ConfigManager implements IManager {
   private static final ConfigNodeConfig CONF = ConfigNodeDescriptor.getInstance().getConf();
   private static final CommonConfig COMMON_CONF = CommonDescriptor.getInstance().getConfig();
 
+  private final ClusterManager clusterManager;
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java:207]: move to line 212

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java [hunk 3]
```diff
@@ -272,6 +277,7 @@ public ConfigManager() throws IOException {
             quotaInfo);
     this.stateMachine = new ConfigRegionStateMachine(this, executor);
 
+    this.clusterManager = new ClusterManager(this, clusterInfo);
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java:280]: move to line 283

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/IoTDBShutdownHook.java [hunk 4]
```diff
@@ -55,6 +55,7 @@ public IoTDBShutdownHook() {
 
   @Override
   public void run() {
+    logger.info("IoTDB exiting...");
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/IoTDBShutdownHook.java:58]: DataNode exiting

**Review by OneSizeFitsQuorum — COMMENTED** [review]
ADD an IT

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/consensus/statemachine/ConfigRegionStateMachine.java (hunk 1 above)
Comment (liyuheng55555) [reply iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/consensus/statemachine/ConfigRegionStateMachine.java:239]: ok

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java (hunk 2 above)
Comment (liyuheng55555) [reply iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java:207]: done

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java (hunk 3 above)
Comment (liyuheng55555) [reply iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java:280]: done

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      long startTime = System.currentTimeMillis();
+      while (System.currentTimeMillis() - startTime < maxTestTime) {
+        try {
+          TGetClusterIdResp resp = client.getClusterId();
+          if (TSStatusCode.SUCCESS_STATUS.getStatusCode() == resp.getStatus().getCode()) {
+            return;
+          }
+        } catch (TException ignore) {
+
+        }
```
Comment (CRZbulabula) [inline integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java:88]: The error exception should be logged
```suggestion
        } catch (TException e) {
          logger.error(e);
        }
```

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+      }
+      String errorMessage = String.format("Cluster ID failed to generate in %d ms.", maxTestTime);
+      Assert.fail(errorMessage);
+    } catch (ClientManagerException | IOException | InterruptedException e) {
+      throw new RuntimeException(e);
```
Comment (CRZbulabula) [inline integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java:93]: No need to throw RuntimeException. Catch and log is enough

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/write/confignode/UpdateClusterIdPlan.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+import org.apache.iotdb.confignode.consensus.request.ConfigPhysicalPlan;
+import org.apache.iotdb.confignode.consensus.request.ConfigPhysicalPlanType;
+import org.apache.iotdb.tsfile.utils.ReadWriteIOUtils;
+
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Objects;
+
+public class UpdateClusterIdPlan extends ConfigPhysicalPlan {
```
Comment (CRZbulabula) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/write/confignode/UpdateClusterIdPlan.java:31]: Please add serialize and deserialize for UpdateClusterIdPlan

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+import org.junit.runner.RunWith;
+import org.slf4j.Logger;
+
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+
+@RunWith(IoTDBTestRunner.class)
+@Category({ClusterIT.class})
+public class IoTDBClusterStartIT {
+  private static final Logger logger = IoTDBTestLogger.logger;
```
Comment (CRZbulabula) [inline integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java:47]: Why not use LoggerFactory.getLogger here?

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/ClusterInfo.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.UUID;
+
+public class ClusterInfo implements SnapshotProcessor {
```
Comment (CRZbulabula) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/ClusterInfo.java:39]: Please add serialize and deserialize test for ClusterInfo

**Review by CRZbulabula — CHANGES_REQUESTED** [review]
PTAL. By the way, it's better to test version updation for this PR

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+  public void clusterIdTest() {
+    try (SyncConfigNodeIServiceClient client =
+        (SyncConfigNodeIServiceClient) EnvFactory.getEnv().getLeaderConfigNodeConnection()) {
+      long startTime = System.currentTimeMillis();
+      while (System.currentTimeMillis() - startTime < maxTestTime) {
+        try {
+          TGetClusterIdResp resp = client.getClusterId();
+          if (TSStatusCode.SUCCESS_STATUS.getStatusCode() == resp.getStatus().getCode()) {
+            return;
+          }
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java:85]: omg sorry I forgot

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java (hunk 6 above)
Comment (liyuheng55555) [reply integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java:93]: Sure, now throw them to upper level

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java (hunk 8 above)
Comment (liyuheng55555) [reply integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java:47]: To follow IoTDBClusterRestartIT :P

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/ClusterInfo.java (hunk 9 above)
Comment (liyuheng55555) [reply iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/ClusterInfo.java:39]: good point !

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/write/confignode/UpdateClusterIdPlan.java (hunk 7 above)
Comment (liyuheng55555) [reply iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/write/confignode/UpdateClusterIdPlan.java:31]: sure

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java (hunk 5 above)
Comment (liyuheng55555) [reply integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java:88]: indeed

**Comment by liyuheng55555** [comment]
> PTAL. By the way, it's better to test version updation for this PR

Checked, it's work.

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java (hunk 8 above)
Comment (CRZbulabula) [reply integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java:47]: But normally we use LoggerFactory.getLogger

**Review by CRZbulabula — COMMENTED** [review]
PTAL

**Review by CRZbulabula — APPROVED** [review]
LGTM!

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+  @Before
+  public void setUp() {
+    EnvFactory.getEnv()
+        .getConfig()
+        .getCommonConfig()
+        .setConfigNodeConsensusProtocolClass(ratisConsensusProtocolClass)
+        .setSchemaRegionConsensusProtocolClass(ratisConsensusProtocolClass)
+        .setDataRegionConsensusProtocolClass(ratisConsensusProtocolClass)
+        .setConfigNodeConsensusProtocolClass(ConsensusFactory.RATIS_CONSENSUS);
```
Comment (OneSizeFitsQuorum) [inline integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java:62]: remove duplicated

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+  @Test
+  public void clusterIdTest() throws ClientManagerException, IOException, InterruptedException {
+    final long maxTestTime = TimeUnit.SECONDS.toMillis(30);
+    final long testInterval = TimeUnit.SECONDS.toMillis(1);
+    try (SyncConfigNodeIServiceClient client =
+        (SyncConfigNodeIServiceClient) EnvFactory.getEnv().getLeaderConfigNodeConnection()) {
+      long startTime = System.currentTimeMillis();
+      while (System.currentTimeMillis() - startTime < maxTestTime) {
+        try {
+          TGetClusterIdResp resp = client.getClusterId();
```
Comment (OneSizeFitsQuorum) [inline integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java:81]: Assert.notEmpty

**Review by OneSizeFitsQuorum — COMMENTED** [review]
PTAL

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java (hunk 11 above)
Comment (liyuheng55555) [reply integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java:62]: removed

File: integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java (hunk 12 above)
Comment (liyuheng55555) [reply integration-test/src/test/java/org/apache/iotdb/confignode/it/cluster/IoTDBClusterStartIT.java:78]: very rigourous

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Pull request merged**